app/pages/reservation/reservation-phases/ReservationPhases.js

```diff
@@ -1,36 +1,49 @@
 import PropTypes from 'prop-types';
 import React from 'react';
-import indexOf from 'lodash/indexOf';
 
+import NumericalProgressSteps from '../../../shared/progress-steps/NumericalProgressSteps';
 import injectT from '../../../i18n/injectT';
-import ReservationPhase from './ReservationPhase';
+import { hasProducts } from '../../../utils/resourceUtils';
 
 ReservationPhases.propTypes = {
   currentPhase: PropTypes.string.isRequired,
+  resource: PropTypes.object,
   isEditing: PropTypes.bool,
   t: PropTypes.func.isRequired,
 };
 
-function ReservationPhases({ currentPhase, isEditing, t }) {
-  const phases = ['information', 'confirmation'];
+const phases = {
+  information: 'ReservationPhase.informationTitle',
+  confirmation: 'ReservationPhase.confirmationTitle',
+  time: 'ReservationPhase.timeTitle',
+  payment: 'ReservationPhase.paymentTitle',
+};
+
+function ReservationPhases({
+  resource,
+  currentPhase,
+  isEditing,
+  t,
+}) {
+  const stepMessageIds = ['information', 'confirmation'];
+
   if (isEditing) {
-    phases.splice(0, 0, 'time');
+    stepMessageIds.splice(0, 0, 'time');
   }
-  const activeIndex = indexOf(phases, currentPhase);
+
+  if (hasProducts(resource)) {
+    stepMessageIds.splice(1, 0, 'payment');
```

Comment:
Wouldn't be better to use a logic that doesn't mutate the original array?

Reply:
`stepMessageIds` is declared inline (not a prop), and it's contents depends on two conditions, `isEditing` and `hasProducts(resource)`. It could be written as
```
  let stepMessageIds;
  if (isEditing && hasProducts(resource)) {
    stepMessageIds = ['time', 'information', 'payment', 'confirmation'];
  } else if (isEditing) {
    stepMessageIds = ['time', 'information', 'confirmation'];
  } else if (hasProducts(resource)) {
    stepMessageIds = ['information', 'payment', 'confirmation'];
  } else {
    stepMessageIds = ['information', 'confirmation'];
  }
```
which does make it clearer what the exact result is, but I don't think that is much better than mutating the array. Adding a new step depending on a state will also be difficult. Do you have any idea how to implement this?

Reply:
If I understood the logic correctly, it could be implemented like: 
```javascript
let stepMessageIds = ['information', 'confirmation'];

if (isEditing) {
  stepMessageIds = [...stepMessageIds, 'time'];
}
if (hasProducts(resource)) {
  stepMessageIds = [...stepMessageIds, 'payment'];
}
```

Reply:
Ah yes, `time` would need to be first which is fine (`stepMessageIds = ['time', ...stepMessageIds]` but payment is tricker since it should be after `information` (I also now see that there is a bug if it is both editing and has products)

Reply:
So now it would be like this:
```
  let stepMessageIds = hasProducts(resource)
    ? ['information', 'payment', 'confirmation']
    : ['information', 'confirmation'];
  if (isEditing) {
    stepMessageIds = ['time', ...stepMessageIds];
  }
```

Reply:
To me, it looks clearer 👍